include/maliput/math/kd_tree.h

```diff
@@ -0,0 +1,265 @@
+      const double d = lhs[i] - rhs[i];
+      dist += d * d;
+    }
+    return dist;
+  }
+};
+
+/// Functor for comparing points according to the given dimension being evaluated at that point.
+/// @tparam Dimension The dimension of the points.
+template <std::size_t Dimension>
+struct NodeCmp {
+  NodeCmp(std::size_t index) : index_(index) {
+    static_assert(Dimension > 0, "Dimension must be greater than 0.");
+    MALIPUT_VALIDATE(index < Dimension, "Index can not be greater than number of dimensions minus one.");
+  }
+  /// Compares two nodes according to the given dimension being evaluated at that point.
+  /// @tparam Node The type of the nodes.
+  ///         It must provide a method get_coordinate() for getting the underlying point.
+  /// @param lhs First node.
+  /// @param rhs Second node.
+  template <typename Node>
+  bool operator()(const Node& lhs, const Node& rhs) const {
+    return lhs.get_coordinate()[index_] < rhs.get_coordinate()[index_];
+  }
+
+  const std::size_t index_{};
+};
+
+}  // namespace details
+
+/// KDTree provides a space-partitioning data structure for organizing points in a k-dimensional space.
+/// The tree is built from a set of points, where each point is a vector of length k.
+/// The tree is built balanced, to guarantee an average of O(log(n)) time for nearest-neighbor queries.
+///
+/// Inspired on https://rosettacode.org/wiki/K-d_tree.
+///
+/// @tparam Coordinate Data type being used, must have:
+/// - operator[] for accessing the value in each dimension.
+/// @tparam Dimension Dimension of the KD-tree.
+/// @tparam Distance A functor used for getting the distance between two coordinates. By default,
+/// details::SquaredDistance is used.
+/// @tparam NodeCmp A functor used for comparing two nodes at certain index/dimension. By default, details::NodeCmp is
+/// used.
+template <typename Coordinate, std::size_t Dimension,
+          typename Distance = details::SquaredDistance<Coordinate, Dimension>,
+          typename NodeCmp = details::NodeCmp<Dimension>>
+class KDTree {
+ public:
+  MALIPUT_NO_COPY_NO_MOVE_NO_ASSIGN(KDTree)
+  static_assert(Dimension > 0, "Dimension must be greater than 0.");
+
+  /// Constructs a KDTree taking a pair of iterators. Adds each
+  /// point in the range [begin, end) to the tree.
+  ///
+  /// @param begin start of range
+  /// @param end end of range
+  /// @tparam Iterator type of the iterator.
+  /// @throws maliput::common::assertion_error When the range is empty.
+  template <typename Iterator>
+  KDTree(Iterator begin, Iterator end) : nodes_(begin, end) {
+    MALIPUT_VALIDATE(!nodes_.empty(), "Empty range");
+    root_ = details::MakeKdTree<Dimension, Node, NodeCmp>(0, nodes_.size(), 0, nodes_);
+  }
+
+  /// Constructs a KDTree taking a vector of points.
+  ///
+  /// @param points Vector of points
+  /// @throws maliput::common::assertion_error When the range is empty.
+  KDTree(const std::vector<Coordinate>& points) {
```

Comment:
I was thinking about this and perhaps we need to just keep the iterator constructor and maybe another constructor to pass in a KDTree(deque<Coordinate>&& points). That way we let users to either copy via iterators or move the data and avoid copies but offering this constructor might not make sense when we have the other and the underlying collection is a deque.

WDYT?

Reply:
I see your point and it is addressed at b55318c
However mind the following:
 - We are passing for example deque<Vector3> and we are storing deque<Node>, where the Node's constructor will get a Vector3. So the "moving" isn't direct.
 - I created a snippet code with this similar case to see if the actual object is being moved and at some point there is a copy: see [coliru snippet code](https://coliru.stacked-crooked.com/a/b00657ec614bb37e). Note how the address of the final point within is different to the original.

Reply:
Solution is OK.

Consider also using std::make_move_iterator() --> https://en.cppreference.com/w/cpp/iterator/make_move_iterator

I think that what is happening is the following:
- The deque is built.
- When emplacing the elements, the move constructor for each Node is called but actually on a new memory position. That simplifies moving each element fron one struct to another but does not mean that new memory will be allocated for that.

It is not exactly the best global solution but one that should allow to reduce the friction of having duplicated elements at the end of the day. The result of this constructor is the data actually moved from one to another.